pkg/global/constant_rate_trace_sampler.go

```diff
@@ -0,0 +1,86 @@
+package global
+
+import (
+	"sync"
+	"time"
+
+	"github.com/buildbarn/bb-storage/pkg/clock"
+
+	"go.opencensus.io/trace"
+)
+
+// ConstantRateTraceSampler uses a token bucket algorithm to sample traces. This is useful as an alternative
+// to probability-based sampling when the service has a low number of requests per second.
+type ConstantRateTraceSampler struct {
+	clock               clock.Clock
+	maxTokens           int64
+	periodDurationNanos int64
+	tokensPerPeriod     int64
+	tokensPerSample     int64
+
+	lock            sync.Mutex
+	availableTokens int64
+	lastRefillTime  time.Time
+	nextRefillTime  time.Time
+}
+
+func max(a, b int64) int64 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// Sample decides whether to sample a trace. The trace.Sampler type is unfortunately a function and
+// not an interface. This `Sample` function should be provided to the OpenCensus tracing library as
+// the `DefaultSampler`, and not the ConstantRateTraceSampler struct.
+func (s *ConstantRateTraceSampler) sample(params trace.SamplingParameters) trace.SamplingDecision {
+	if params.ParentContext.IsSampled() {
```

Comment:
Does this mean that if there's a process in front of the cluster (e.g., an on-prem instance of bb-storage for caching) that reports that a part of the trace is sampled, we always sample without respecting the maximum token count? Is this desirable? May it be abused?

Reply:
> Is this desirable? 

In my view, that question is one for the user. One risk is incomplete traces across the entire life cycle of a request because only certain services choose to log their spans. The benefit is getting complete traces through the system more consistently than if each service made the decision in isolation randomly or at a low rate.

The existing probability-based sampler respects the decision. See https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/trace/sampling.go#L52-L54. 

A more nuanced configuration would be to have a separate sampler that only checks whether the upstream made a trace decision. OpenCensus unfortunately does not have such a sampler; OpenTelemetry does. See https://github.com/open-telemetry/opentelemetry-go/blob/875a25835fcfcddd61e94d87b1a0b14db4d5a0e5/sdk/trace/sampling.go#L157-L162. 

So maybe we should make the parent-sampling part of this PR configurable? The better tracing architecture in my view is to let the user decide what the questions of when to sample at all and when to respect the upstream as separate decisions.

I made the decision here for consistency with the probability-based sampler in OpenCensus, but that is not my preferred choice if consistency is not a requirement.

> May it be abused?

Like anything else, yes. A malicious client could send a `grpc-trace-bin` and now the request will have a tracing context if trace header propagation is enabled if it always respects the parent decision. I've seen multi-service traces across multiple BuildBarn services for a request in the tracing store I'm using, so I suspect something might be enabling that, but I still need to figure out what since I want to disable it if it is enabled on the bb-storage fronting our cluster. (I still see missing spans but that may be occurring for other reasons.)

Do you have any additional context on how BuildBarn's grpc library is setup or not for tracing? 

Any way, I would say it is definitely good practice for any entry point to ignore the `grpc-trace-bin` header and originate the tracing decision itself if it cannot trust its clients.